### PR TITLE
Detect addition and removal of type declarations

### LINKED ISSUE #80: Detect removal and addition of type declarations
The current implementation of our API diffing does not handle every type of item that can be defined within a module.

It will compare items that we support, detecting addition, removal and modification of said items. If it fails to detect any such change, it will default back to comparing the whole module using `Includemod.signatures`.

Our ultimate goal is to support fine grained diffing of each and every signature items.

To get one step closer, we would like to add detection of the addition and removal of type_declaration items.

## How to proceed

The diffing logic and representation is found in [`lib/diff.ml`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/lib/diff.ml).

The content of a module is represented as a list even though the order does not matter there. To work around that we first build a map of its content, indexed by pairs of item type and name. We then proceed to compare the two maps we built from the two versions of the module we are diffing. If both map contain a mapping for a given type/name pair, we compare them to try and detect whether it was modified. If only one of the map contains a binding for a given pair, that means that it was either added or removed, based on which map has the binding.

To detect addition/removal of type declarations, we will have to extend the [`item_type`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/lib/diff.ml#L23) and [`sig_items`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/lib/diff.ml#L24) types which are used to build the map mentioned above. These are then used in the `extract_items` function that builds the map from the module signature.

Finally, the actual diffing logic's entry point is the [`Diff.items`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/lib/diff.ml#L79) function which handles the comparison of the two maps via the merge function (you can read the documentation [here](https://ocaml.org/manual/5.2/api/Map.S.html#VALmerge)).

## Testing

This should be tested at the library level. You can add a new `tests/api-watch/test_diff_type_decl.ml` file where you can write tests for type addition or removal. You can use the [`tests/api-watch/test_diff.ml`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/tests/api-watch/test_diff.ml) or [`tests/api-watch/test_diff_module.ml`](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/tests/api-watch/test_diff_module.ml) files as examples on how to write such tests.

### PR DESCRIPTION
**Pull request by azzsal**
This PR aims to close #80. I will follow TDD. I have updated tests in  [tests/api-watch/test_diff.ml](https://github.com/ocaml-semver/ocaml-api-watch/blob/main/tests/api-watch/test_diff.ml) to detect the addition and removal type declarations.
The PR is still a work in progress and will updated accordingly.

### CODE REVIEW

**Comment by azzsal** [comment]
I have updated the Diff module to include support for addition and removal of type declarations diffing. The tests cannot run since the TextDiff module need to be updated to handle type declarations. Before I go ahead and processed, could you review if my changes in the Diff module are good ?

Thank you!

File: lib/diff.ml
Comment (panglesd) [inline lib/diff.ml:72]: You can make this last case more explicit, instead of using a wildcard. This makes it easier to read what case(s) remains, and lets OCaml's match exhaustiveness check tells you if you have forgotten to treat a case!

**Review by panglesd — COMMENTED** [review]
Looks good to me to continue!

**Comment by azzsal** [comment]
I have updated the rest of Text_diff module and tests to accommodate for type declarations. All tests now pass!

File: lib/text_diff.ml
Comment (panglesd) [inline lib/text_diff.ml:56]: Even though the detection of modified types is not implemented, the printing of modified type is simple enough to be implemented now!

File: lib/text_diff.ml
Comment (panglesd) [inline lib/text_diff.ml:111]:
```suggestion
```

File: tests/api-diff/run.t
Comment (panglesd) [inline tests/api-diff/run.t:1]: Could you also add a test for upcoming modified types?

**Review by panglesd — APPROVED** [review]
Looks ready for merge, with minor comments below. 
You already need to rebase, and I'd prefer to merge #88 and rebase before merging this one.

We'll wait for @NathanReb's review as well.
In the meantime, I think you can work on detection of modified types.

File: tests/api-watch/test_diff_module.ml
Comment (NathanReb) [inline tests/api-watch/test_diff_module.ml:88]: It might be good to change the code a bit here so that the test output still contains an unsupported change.

Once all of the items are handled this part of the test won't be useful anymore but for now, you should be able to simply replace the `type t` with an exception declaration:
```
type exn += Some_exn
```

None of the issues we opened will cover this so it should fix it for other similar PRs as well.

**Review by NathanReb — COMMENTED** [review]
Looks great! I added one comment about one of the changes in the tests.

You might have to rebase on top of the latest changes though before fixing it as #86 has been merged now.

**Comment by azzsal** [comment]
Sorry for the messy commits, but I am not familiar with rebase and more used to merge. The cram tests are not passing for some reason, I will try to fix that.

**Comment by panglesd** [comment]
Indeed, your rebase seems to have gone wrong. There is now more conflicts and more commits!

(I'm personally more used to rebase that's why I suggested it, but you could have done a merge if you prefer!)

Now, my advice would be:
- Create a new branch starting on main,
- cherry-pick the commits you need from the old branch to this new branch, one by one, resolving conflicts as they come
- once everything seems fine, replace the old branch by the new one.

(But maybe there is a better way!)

**Comment by azzsal** [comment]
This can now be safely closed, since all the work is moved to #92. Sorry about the confusion.

**Comment by azzsal** [comment]
This can now be safely closed, since all the work is moved to https://github.com/ocaml-semver/ocaml-api-watch/pull/92. Sorry about the confusion.

**Pull request closed without merging**
